src/prefect/core/task.py

```diff
@@ -368,6 +373,28 @@ def __init__(
             raise ValueError(
                 "A datetime.timedelta `retry_delay` must be provided if max_retries > 0"
             )
+        if retry_on and not max_retries > 0:
+            raise ValueError(
+                "A numer of `max_retries` must be provided if `retry_on` is set."
+            )
+        if retry_on:
+            try:
+                retry_on = set(retry_on)
+            except TypeError as exc:
+                raise TypeError(
+                    "Unexpected type for `retry_on`. "
+                    "Expected an iterable of exception types."
+                ) from exc
+            for v in retry_on:
```

Comment:
For clarity, can we rename `v` to `exc_type`?

Reply:
I could, but it's a little misleading since it's not confirmed to be an exception type yet. At this point, all we know is that it's some value. Thoughts?